src/data/finsupp/basic.lean

```diff
@@ -1069,6 +1092,19 @@ lemma prod_add_index' [add_comm_monoid M] [comm_monoid N] {f g : α →₀ M}
     f.prod (λ a b, h a (multiplicative.of_add b)) * g.prod (λ a b, h a (multiplicative.of_add b)) :=
 prod_add_index (λ a, (h a).map_one) (λ a, (h a).map_mul)
 
+lemma sum_emb_domain_index [add_comm_monoid M] [add_comm_monoid N]
+  (f : α ↪ β) (v : α →₀ M) (g : β → M →+ N) :
+  (v.emb_domain f).sum (λ b m, g b m) = v.sum (λ a m, g (f a) m) :=
+begin
+  apply induction_linear v,
+  { simp, },
+  { intros v₁ v₂ h₁ h₂,
+    rw [emb_domain_add, sum_add_index, sum_add_index],
+    { simp [h₁, h₂], },
+    all_goals { simp, }, },
+  { intros, simp, },
+end
```

Comment:
You don't use the additive structure on `g` at all here. This generalizes to:
```
@[to_additive]
lemma prod_emb_domain_index [comm_monoid M] [comm_monoid N]
  (f : α ↪ β) (v : α →₀ M) (g : β → M → N) :
  (v.emb_domain f).prod (λ b m, g b m) = v.prod (λ a m, g (f a) m) :=
by rw [emb_domain_eq_map_domain, prod_map_domain_index_inj f.injective]
```

Reply:
Much better, thank you!